### DISCUSSION TITLE
@stack / @push / @prepend don't work as expected

### DISCUSSION THREAD

**Discussion by smares** [laravel/framework#36584] [Q&A]
In `template.blade.php`:

```
@push('scripts')
    <script src="{{ mix('js/manifest.js') }}"></script>
    <script src="{{ mix('js/vendor.js') }}"></script>
    <script src="{{ mix('js/app.js') }}"></script>
@endpush
<!DOCTYPE html>
<html lang="@yield('language', str_replace('_', '-', app()->getLocale()))">
<head>
<meta charset="utf-8" />
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1.0" />

<title>@yield('title', config('app.name'))</title>

<link href="{{ mix('css/app.css') }}" rel="stylesheet" type="text/css" />
</head>
<body>
@yield('content')
@stack('scripts')
</body>
</html>
```

In `page.blade.php`:

```
@extends('base.three-columns')

@section('content')
...
@ensection

@push('scripts')
    <script>123</script>
@endpush
```

When rendering the page, I would expect `123` to come after the scripts pushed by the template, but `123` comes first. Same happens if I have a Blade component that pushes something to `scripts`.

What I am actually trying to achieve is define the site-wide scripts somewhere and then use `@push` and `@prepend` to load scripts before or after the site-wide scripts. As a work-around, I will be using two stacks, something like this in the `template.blade.php`:

```
...
@stack('scripts-before')
<script src="{{ mix('js/manifest.js') }}"></script>
<script src="{{ mix('js/vendor.js') }}"></script>
<script src="{{ mix('js/app.js') }}"></script>
@stack('scripts-after')
</body>
...
```

Am I doing something wrong or is this a bug?

**Comment by rodrigopedra** [comment]
> When rendering the page, I would expect 123 to come after the scripts pushed by the template, but 123 comes first. 

Why would you expect that?

1. A route returns `view('page')`
2. `page.blade.php` gets parsed, `@push` pushes `123` to stack (it is the first thing pushed onto it)
3. `@extends` is found, while parsing, tries to fetch `'base.three-columns'`
4. `base/three-columns.blade.php` is parsed I assume it extends the `template` on your first snippet
5. `template.blade.php` is parsed, `@push` contents is pushed into the stack (after `123`)
6. `@stack` is found, stack is rendered, first `123` from `page.blade.php` then scripts from `template.blade.php`

Maybe you are assuming the master template would be executed before its children templates, that is not how it works.

More details on:

https://github.com/laravel/framework/blob/72ea328b456ea570f8823c69f511583aa6234170/src/Illuminate/View/Concerns/ManagesStacks.php

**Reply by smares** [reply]
Is it possible at all to achieve what I am trying to do without using two different stacks?

**Reply by rodrigopedra** [reply]
Sure, using the two different stacks before and after the script tags should do the trick

**Reply by rodrigopedra** [reply]
Sorry I misread your last message, didn't see the "without", it was late here when I answered.

Updating my answer then: I don't think you could achieve that without using two different stacks.
